infer/src/backend/dom.ml

```diff
@@ -2169,55 +2169,145 @@ let pathset_join pname tenv (pset1: Paths.PathSet.t) (pset2: Paths.PathSet.t)
+    let maxclique zs = (* is zs a maxclique for the subgraph induced by 0..y? *)
+      let rec loop u vs = u > y || (match u, vs with
+        | u, v :: vs when Int.equal u v -> loop (u + 1) vs
+        | u, vs -> not (List.for_all zs ~f:(edge u)) && loop (u + 1) vs) in
+      loop 0 zs in
+    let fillup zs = (* greedily add vertices from y+1..n-1 to clique *)
+      let compatible z zs = List.for_all zs ~f:(edge z) in
+      filter_to n compatible (List.rev zs) (y + 1) in
+    if maxclique zs then S.add (fillup zs) queue else queue in
+  let queue = List.fold ~init:queue ~f:do_y ys in
+  (xs, (queue, n, edge))
+
+(* Find maximal subsets of [xs] that can be combined with a binary operator
+[plus] that is associative, commutative, but not necessarily defined
+everywhere.
+
+Let's call set S={x1,...,xn} *consistent* when ΣS=x1+...+xn is defined.  When
+x1,...,xn are preconditions and [plus] is the meet operator, it happens often
+(but not always) that pairwise consistency implies set consistency.  Therefore,
+we will use an algorithm that works fast in this case, and is incomplete
+otherwise.
+
+Suppose pairwise consistency implies set consistency.  Define a graph whose
+vertices are the [xs] and has an (undirected) edge xy when x+y is defined. In
+terms of this graph, the problem is to generate all maximal cliques. This can
+be done with polynomial delay using the JYP algorithm:
+  Johnson, Yannakakkis, Papadimitriou,
+  On Generating All Maximal Independent Sets, 1987
+A maximal clique S found by this algorithm will be pairwise consistent, and
+often but not always consistent. The current implementation simply filters
+out inconsistent maximal cliques.
+
+The [timeout] limits the number of maxcliques to be enumerated, because they
+can be exponentially many.
+*)
+let maximal_combine ~(timeout : int) ~(init : 'a) ~(plus : 'a -> 'a -> 'a option) (xs : 'a list)  : 'a list =
+  let n = List.length xs in
+  let by_idx =
+    let xs = Array.of_list xs in
+    Array.get xs in
+  let lost = ref 0 in (* counts inconsistent maximal cliques *)
+  let exception Undefined in
+  let plus_exc x y = match plus x y with
+    | None -> raise Undefined
+    | Some z -> z in
+  let pairs = (* first, build graph with pairwise consistency *)
+    let h = Hashtbl.create (n * n) in
+    for i = 0 to n - 1 do
+      for j = i + 1 to n - 1 do
+        (try
+          let z = plus_exc (by_idx i) (by_idx j) in
+          Hashtbl.add h (i, j) z;
+          Hashtbl.add h (j, i) z
+        with Undefined -> ())
+      done
+    done; h in
+  let edge i j = Hashtbl.mem pairs (i, j) in
+  let maxclique_count = ref 0 in
+  let exception Timeout in
+  let result = ref [] in
+  let rec loop g =
+    SymOp.pay ();
+    incr maxclique_count;
+    if !maxclique_count >= timeout then raise Timeout;
+    let xs, ng = next_maxclique g in (* xs is a pairwise consistent set *)
+    let zs = List.map ~f:by_idx xs in
+    (try result :=  (List.fold ~init ~f:plus_exc zs) :: !result
```

Comment:
I am wondering if a cache of prefixes would make sense

Reply:
seems to: cache hit quite high
(also, could be used for the vertex cover, if needed later)